README.md
# Better Totem of Undying [FABRIC]

[![Status](https://img.shields.io/badge/development%20status-Active-brightgreen?style=for-the-badge)]()
[![Discord](https://img.shields.io/discord/834250417130831903?color=blue&label=Discord&logo=discord&style=for-the-badge)](https://discord.gg/2PvYZfjWDf)
[![CurseForge](https://img.shields.io/static/v1?label=Curseforge&message=Fabric&color=orange&labelColor=grey&style=for-the-badge&logo=curseforge&logoColor=orange)](https://www.curseforge.com/minecraft/mc-mods/better-totem-of-undying-fabric)
[![Modrinth](https://img.shields.io/static/v1?label=Modrinth&message=Fabric&color=dark_green&labelColor=grey&style=for-the-badge&logo=Modrinth&logoColor=dark_green)](https://modrinth.com/mod/better-totem-of-undying-forge)

A highly configurable mod that improves Minecraft Totem of Undying by adding new features to it. You can blacklist dimensions or even structures where you don't want totem to work. The mod also has compatibility with Trinkets mod.

## Features 

- Totem will knockback mobs when they are too close.
- Totem will save you when falling into the void, teleporting you safely to the surface.
- Totem will destroy blocks if you are suffocating inside them. (If there is a block you don't want to be destroyed, you can add it to the "totem_cant_break" tag).
- Totem will increase your hunger bar to 8 if it's <= 6 (When you can't run). Giving you enough energy to run from danger and/or find food.
- Totem will also apply Water Breathing effect. However, both fire resistance and water breathing effects will only be applied when needed. This means you are going to receive the effects only if you are on fire or in water, respectively.
- You can use the totem from charm slot if Trinkets mod is installed.
- You can use the totem from your inventory (Disabled by default).
- You can add a cooldown to the totem (Disabled by default).
- You can blacklist dimensions and structures where you don't want totem to work.

## Gifs

<details><summary>Click here (GIFS)</summary>
<p>

![Totem Knockback](https://github.com/CerbonXD/BetterTotemOfUndying/blob/master/.github/Totem%20Knockback.gif)

![Totem Teleport Out Of Void](https://github.com/CerbonXD/BetterTotemOfUndying/blob/master/.github/Totem%20Teleport%20Out%20Of%20Void.gif)

![Destroy Blocks When Suffocating](https://github.com/CerbonXD/BetterTotemOfUndying/blob/master/.github/Destroy%20Blocks%20When%20Suffocating.gif)

</p>
</details>

## Config File 

The config file is partitioned in four sections:

- **Minecraft Default Totem Features:** Here is where you can configure things Minecraft does by default with totem, such as giving you regeneration and absorption effects.
- **New Totem Features:** Here is where you can configure all new totem features added by the mod.
- **Trinkets:** Here is where you can configure things related to trinkets integration.
- **Blacklists:** Here is where you can blacklist dimensions or structures.

You need to have [cloth-api](https://www.curseforge.com/minecraft/mc-mods/cloth-config) to edit configs.

<details><summary>Click here (Config)</summary>
<p>

```toml

["Minecraft Default Totem Features"]
	#This value sets the health Totem of Undying will give to the entity upon use. DEFAULT: 1
	#Range: 0 ~ 20
	"Set Health" = 1
	#When Totem of Undying is used it removes all previous effects you had. If set to false, it will keep all the effects you had before using the totem. DEFAULT: TRUE
	"Remove All Effects" = true

	["Minecraft Default Totem Features".Effects]
		#If false Totem of Undying will not give you fire resistance effect. DEFAULT: TRUE
		"Enable Fire Resistance" = true
		#Sets the duration of the fire resistance effect in ticks. DEFAULT: 800
		"Fire Resistance Duration" = 800
		#If false Totem of Undying will not give you regeneration effect. DEFAULT: TRUE
		"Enable Regeneration" = true
		#Sets the duration of the regeneration effect in ticks. DEFAULT: 900
		"Regeneration Duration" = 900
		#Sets the amplifier of the regeneration effect. DEFAULT: 1
		#Range: 0 ~ 255
		"Regeneration Amplifier" = 1
		#If false Totem of Undying will not give you absorption effect. DEFAULT: TRUE
		"Enable Absorption" = true
		#Sets the duration of the absorption effect in ticks. DEFAULT: 100
		"Absorption Duration" = 100
		#Sets the amplifier of the absorption effect. DEFAULT: 1
		#Range: 0 ~ 255
		"Absorption Amplifier" = 1

["New Totem Features"]
	#If true you will be able to use the Totem of Undying from your inventory. DEFAULT: FALSE
	"Use Totem From Inventory" = false

	["New Totem Features".Effects]
		#Apply fire resistance and water breathing effects only when needed, meaning you are going to receive the effects only if you are on fire or in water respectively. DEFAULT: TRUE
		"Apply Effects Only When Needed" = true
		#If false Totem of Undying will not give you water breathing effect. DEFAULT: TRUE
		"Enable Water Breathing" = true
		#Sets the duration of the water breathing effect in ticks. DEFAULT: 800
		"Water Breathing Duration" = 800
		#If false Totem of Undying will not give you slow falling effect. DEFAULT: TRUE
		"Enable Slow Falling" = true
		#Sets the duration of the slow falling effect in ticks. DEFAULT: 600
		"Slow Falling Duration" = 600

	["New Totem Features"."Increase Food Level"]
		#If false Totem of Undying will not increase your food level. DEFAULT: TRUE
		"Increase Food Level" = true
		#Sets the minimum food level needed to Totem of Undying increase food level. DEFAULT: <= 6
		#Range: 0 ~ 20
		"Minimum Food Level" = 6
		#Sets the food level that Totem of Undying will give upon use. DEFAULT: 8
		#Range: 0 ~ 20
		"Set Food Level" = 8

	["New Totem Features"."Destroy Blocks When Suffocating or Fully Frozen"]
		#If false Totem of Undying will not break blocks when you are suffocating. DEFAULT: TRUE
		"Destroy Blocks When Suffocating" = true
		#If false Totem of Undying will not break the powder snow when you die fully frozen. DEFAULT: TRUE
		"Destroy Powder Snow When Fully Frozen" = true

	["New Totem Features"."Knockback Mobs Away"]
		#If false Totem of Undying will not knockback mobs away. DEFAULT: TRUE
		"Knockback Mobs Away" = true
		#Sets the radius where entities needs to be for the totem knockback them. DEFAULT: 3.0
		"Knockback Radius" = 3.0
		#Sets the strength of the knockback. DEFAULT: 2.5
		"Knockback Strength" = 2.5

	["New Totem Features"."Teleport Out of Void"]
		#If false Totem of Undying will not save you from dying in the void. DEFAULT: TRUE
		"Teleport Out of Void" = true
		#If totem can't find a available position to teleport you back it will teleport you to the world's max build height plus this offset. DEFAULT:64
		#Range: 0 ~ 1024
		"Teleport Height Offset" = 64

	["New Totem Features"."Add Cooldown"]
		#If true Totem of Undying will receive a cooldown after being used and you will not be able to use it again during this period. DEFAULT: FALSE
		"Add Cooldown" = false
		#Sets the cooldown duration in ticks. DEFAULT: 200
		Cooldown = 200

[Curios]
	#If false you will not be able to use Totem of Undying from charm slot (Curios mod must be installed). DEFAULT: TRUE
	"Use Totem From Charm Slot" = true
	#If false Totem of Undying will not be displayed on the chest when worn in the curios charm slot (Curios mod must be installed). DEFAULT: TRUE
	"Display Totem on Chest" = true

[Blacklists]
	#You can put here dimensions where you don't want Totem of Undying to work. Example: "minecraft:overworld", "mod_id:dimension_id" DEFAULT: Nothing
	"Blacklisted Dimensions" = []
	#You can put here structures where you don't want Totem of Undying to work. Example: "minecraft:desert_pyramid", "mod_id:structure_id" DEFAULT: Nothing
	"Blacklisted Structures" = []
```

</p>
</details>

## Support

Please report all bugs, issues, and feature requests to the [issue tracker](https://github.com/CerbonXD/BetterTotemOfUndying-FABRIC/issues) or use my [discord server](https://discord.gg/2PvYZfjWDf) and report at the appropriate channel.

Please read the [FAQ](https://github.com/CerbonXD/BetterTotemOfUndying-FABRIC/blob/master/FAQ.md) before reporting a bug, issue or feature request. 

## Donations 

[![ko-fi](https://ko-fi.com/img/githubbutton_sm.svg)](https://ko-fi.com/E1E8MI65L)

## Affiliates

[![BisectHosting](https://www.bisecthosting.com/partners/custom-banners/1a9a55fc-f1c0-4b40-b07c-3774bc557f93.webp)](https://bisecthosting.com/Cerbon)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.1.1] - 2023-10-09

### Fixed

- Fix crash during world creation
- Fix wrong file name

## [1.1.0] - 2023-09-09

### Added

 - You can now set custom effects when the totem is activated. See the [FAQ](https://github.com/CERBON-MODS/CERBONs-Better-Totem-of-Undying-FABRIC/blob/master/FAQ.md) to understand how to use it

## [1.0.1] - 2023-07-26

### Changed

- Improved code to ensure better compatibility between other mods and to follow Mixin standards

## [1.0.0] - 2023-07-12

- Initial release


src/main/java/com/cerbon/better_totem_of_undying/config/BTUConfigs.java
package com.cerbon.better_totem_of_undying.config;

import com.cerbon.better_totem_of_undying.util.BTUConstants;
import me.shedaniel.autoconfig.ConfigData;
import me.shedaniel.autoconfig.annotation.Config;
import me.shedaniel.autoconfig.annotation.ConfigEntry;
import me.shedaniel.cloth.clothconfig.shadowed.blue.endless.jankson.Comment;

import java.util.ArrayList;
import java.util.List;

@Config(name = BTUConstants.MOD_ID)
public class BTUConfigs implements ConfigData {

    @ConfigEntry.Category("minecraft_default_totem_features")
    @Comment("This value sets the health Totem of Undying will give to the entity upon use. DEFAULT: 1")
    @ConfigEntry.BoundedDiscrete(min = 1, max = 20)
    public int SET_HEALTH = 1;

    @ConfigEntry.Category("minecraft_default_totem_features")
    @Comment("When Totem of Undying is used it removes all previous effects you had. If set to false, it will keep all the effects you had before using the totem. DEFAULT: TRUE")
    public boolean REMOVE_ALL_EFFECTS = true;

    @ConfigEntry.Category("minecraft_default_totem_features")
    @Comment("If false Totem of Undying will not give you fire resistance effect. DEFAULT: TRUE")
    public boolean ENABLE_FIRE_RESISTANCE = true;

    @ConfigEntry.Category("minecraft_default_totem_features")
    @Comment("Sets the duration of the fire resistance effect in ticks. DEFAULT: 800")
    public int FIRE_RESISTANCE_DURATION = 800;

    @ConfigEntry.Category("minecraft_default_totem_features")
    @Comment("If false Totem of Undying will not give you regeneration effect. DEFAULT: TRUE")
    public boolean ENABLE_REGENERATION = true;

    @ConfigEntry.Category("minecraft_default_totem_features")
    @Comment("Sets the duration of the regeneration effect in ticks. DEFAULT: 900")
    public int REGENERATION_DURATION = 900;

    @ConfigEntry.Category("minecraft_default_totem_features")
    @Comment("Sets the amplifier of the regeneration effect. DEFAULT: 1")
    @ConfigEntry.BoundedDiscrete(min = 0, max = 255)
    public int REGENERATION_AMPLIFIER = 1;

    @ConfigEntry.Category("minecraft_default_totem_features")
    @Comment("If false Totem of Undying will not give you absorption effect. DEFAULT: TRUE")
    public boolean ENABLE_ABSORPTION = true;

    @ConfigEntry.Category("minecraft_default_totem_features")
    @Comment("Sets the duration of the absorption effect in ticks. DEFAULT: 100")
    public int ABSORPTION_DURATION  = 100;

    @ConfigEntry.Category("minecraft_default_totem_features")
    @Comment("Sets the amplifier of the absorption effect. DEFAULT: 1")
    @ConfigEntry.BoundedDiscrete(min = 0, max = 255)
    public int ABSORPTION_AMPLIFIER  = 1;


    @ConfigEntry.Category("new_totem_features")
    @Comment("If true you will be able to use the Totem of Undying from your inventory. DEFAULT: FALSE")
    public boolean USE_TOTEM_FROM_INVENTORY = false;

    @ConfigEntry.Category("new_totem_features")
    @Comment("Apply fire resistance and water breathing effects only when needed, meaning you are going to receive the effects only if you are on fire or in water respectively. DEFAULT: TRUE")
    public boolean APPLY_EFFECTS_ONLY_WHEN_NEEDED = true;

    @ConfigEntry.Category("new_totem_features")
    @Comment("If false Totem of Undying will not give you water breathing effect. DEFAULT: TRUE")
    public boolean ENABLE_WATER_BREATHING = true;

    @ConfigEntry.Category("new_totem_features")
    @Comment("Sets the duration of the water breathing effect in ticks. DEFAULT: 800")
    public int WATER_BREATHING_DURATION = 800;

    @ConfigEntry.Category("new_totem_features")
    @Comment("If false Totem of Undying will not give you slow falling effect. DEFAULT: TRUE")
    public boolean ENABLE_SLOW_FALLING = true;

    @ConfigEntry.Category("new_totem_features")
    @Comment("Sets the duration of the slow falling effect in ticks. DEFAULT: 600")
    public int SLOW_FALLING_DURATION = 600;

    @ConfigEntry.Category("new_totem_features")
    @ConfigEntry.Gui.Excluded
    @Comment("Use this list to add custom effects and specify when they should be triggered. Within this list, you should provide another list using curly braces, with the first parameter being the damage type, the second parameter being the effect that will be granted, the third parameter the duration of the effect in ticks, and the fourth parameter the effect amplifier. If you want the effect to always be granted, use the damage type as \"any\". Example: [[\"minecraft:out_of_world\", \"minecraft:blindness\", \"800\", \"0\"]]. In this example, if the player dies in the void, the totem will grant the level 1 blindness effect with a duration of 40 seconds. DEFAULT: Nothing")
    public List<List<String>> CUSTOM_EFFECTS = List.of();

    @ConfigEntry.Category("new_totem_features")
    @Comment("If false Totem of Undying will not increase your food level. DEFAULT: TRUE")
    public boolean ENABLE_INCREASE_FOOD_LEVEL = true;

    @ConfigEntry.Category("new_totem_features")
    @Comment("Sets the minimum food level needed to Totem of Undying increase food level. DEFAULT: <= 6")
    public int MINIMUM_FOOD_LEVEL = 6;

    @ConfigEntry.Category("new_totem_features")
    @Comment("Sets the food level that Totem of Undying will give upon use. DEFAULT: 8")
    public int SET_FOOD_LEVEL = 8;

    @ConfigEntry.Category("new_totem_features")
    @Comment("If false Totem of Undying will not break blocks when you are suffocating. DEFAULT: TRUE")
    public boolean DESTROY_BLOCKS_WHEN_SUFFOCATING = true;

    @ConfigEntry.Category("new_totem_features")
    @Comment("If false Totem of Undying will not break the powder snow when you die fully frozen. DEFAULT: TRUE")
    public boolean DESTROY_POWDER_SNOW_WHEN_FULLY_FROZEN = true;

    @ConfigEntry.Category("new_totem_features")
    @Comment("If false Totem of Undying will not knockback mobs away. DEFAULT: TRUE")
    public boolean KNOCKBACK_MOBS_AWAY = true;

    @ConfigEntry.Category("new_totem_features")
    @Comment("Sets the radius where entities needs to be for the totem knockback them. DEFAULT: 3.0")
    public double KNOCKBACK_RADIUS = 3.0D;

    @ConfigEntry.Category("new_totem_features")
    @Comment("Sets the strength of the knockback. DEFAULT: 2.5")
    public double KNOCKBACK_STRENGTH = 2.5D;

    @ConfigEntry.Category("new_totem_features")
    @Comment("If false Totem of Undying will not save you from dying in the void. DEFAULT: TRUE")
    public boolean TELEPORT_OUT_OF_VOID = true;

    @ConfigEntry.Category("new_totem_features")
    @Comment("If totem can't find a available position to teleport you back it will teleport you to the world's max build height plus this offset. DEFAULT:64")
    @ConfigEntry.BoundedDiscrete(min = 0, max = 1024)
    public int TELEPORT_HEIGHT_OFFSET = 64;

    @ConfigEntry.Category("new_totem_features")
    @Comment("If true Totem of Undying will receive a cooldown after being used and you will not be able to use it again during this period. DEFAULT: FALSE")
    public boolean ADD_COOLDOWN = false;

    @ConfigEntry.Category("new_totem_features")
    @Comment("Sets the cooldown duration in ticks. DEFAULT: 200")
    public int COOLDOWN = 200;


    @ConfigEntry.Category("trinkets")
    @Comment("If false you will not be able to use Totem of Undying from trinkets slot (Trinkets mod must be installed). DEFAULT: TRUE")
    public boolean USE_TOTEM_FROM_TRINKETS_SLOT = true;

    @ConfigEntry.Category("trinkets")
    @Comment("If false Totem of Undying will not be displayed on the chest when worn in the trinkets charm slot (Trinkets mod must be installed). DEFAULT: TRUE")
    public boolean DISPLAY_TOTEM_ON_CHEST = true;


    @ConfigEntry.Category("blacklists")
    @Comment("You can put here dimensions where you don't want Totem of Undying to work. Example: \"minecraft:overworld\", \"mod_id:dimension_id\" DEFAULT: Nothing")
    public List<String> BLACKLISTED_DIMENSIONS = new ArrayList<>();

    @ConfigEntry.Category("blacklists")
    @Comment("You can put here structures where you don't want Totem of Undying to work. Example: \"minecraft:desert_pyramid\", \"mod_id:structure_id\" DEFAULT: Nothing")
    public List<String> BLACKLISTED_STRUCTURES = new ArrayList<>();
}
